### DISCUSSION TITLE
Im Stuck! Wally Hangs with:    CMD8: Failed, 0x00F00001AA Failed at: 0x00000000000017B2 on Arty-A7 100

### DISCUSSION THREAD

**Discussion by PietDelaney** [openhwgroup/cvw#1213] [Q&A] [answered]
![Screenshot from 2024-12-30 21-34-07](https://github.com/user-attachments/assets/ec99c7dc-0220-4f2f-a6f3-7bf1a61003ae)

Welcome to minicom 2.8

OPTIONS: I18n                                                                
Port /dev/ttyUSB1, 21:12:38                                                  
                                                                             
Press CTRL-A Z for help on special keys                                      
                                                                                  
 █▀█        █▀█        █▀█        █▀▀                 █ █                         
 █          █ █        █▄▀        █▄▄       ▄▄▄       █ █                         
 █▄█        █▄█        █ █        █▄▄                 ▀▄▀                         
 ____          ____  ____      ___      ___   ____    ___                         
 \   \        /   / /    \    |   |    |   |  \   \  /  /                    
  \   \  __  /   / /      \   |   |    |   |   \   \/  /                     
   \   \/  \/   / /   /\   \  |   |    |   |    \     /                      
    \          / /   ____   \ |   |___ |   |___  |   |                       
     \___/\___/ /___/    \___\|_______||_______| |___|                       
                                                                             
[0.05342] Initializing SPI Controller.                                       
[0.05703] Initializing SD Card in SPI mode.                                  
[0.06125] CMD0: Success, r = 0x01                                            
[0.06435] CMD8: Failed, 0x00F00001AA
[0.06788] Failed at: 0x00000000000017B2

I used minicom to access the console via the USB port.

Anything obviously wrong to anyone?

Here's the vivado screen:

![Vivado_Screenshot from 2024-12-30 21-38-30](https://github.com/user-attachments/assets/4581c94a-5e31-4be8-9aa3-73a04615370e)

-piet

**Comment by PietDelaney** [comment]
![Console_Screenshot from 2024-12-30 21-44-37](https://github.com/user-attachments/assets/21330831-692b-4176-af77-97cda0298009)

**Reply by JacobPease** [reply]
CMD8 returns a 5 byte response and it should return the value `0x01000001AA`. Does it consistently give you the value `0x00F00001AA` when resetting the Arty? I don't think you changed anything, but just in case, did you happen to make any changes to the clock frequencies before building the zsbl?

**Reply by PietDelaney** [reply]
Thanks Jacob:

The console log is the some on multiple times.

I don't think I change a clock frequency.

What is zsbi?

What do 'CMD0" and 'CMD8" mean/imply?

-piet

**Reply by JacobPease** [reply]
`CMD0` and `CMD8` are SD card protocol commands. `CMD0` resets the card into an idle state. It's single byte response tells us if it successfully transitioned into the idle state. During the sending of `CMD0` if we `assert` the `MOSI` line, we can put the card in SPI mode instead of SDIO mode. SPI mode has a slightly different protocol from SDIO mode.

`CMD8` writes to an internal control register on the card. It then responds with a portion of the value you sent to confirm that the write was successful. In this case the first 2 bytes of the response are corrupted.

We were having issues like this about 2-3 months ago. It was because the SPI controller, at certain clock speeds, wouldn't correctly read bytes into it's receive FIFO, but we fixed these issues and have been booting Linux consistently for the past 2-3 months. Have you pulled the latest changes from the main branch since then?

**Comment by rosethompson** [comment]
Hi Piet,

We've had this problem before with past versions of the zero stage bootloader which copies the flash card to DRAM.  Can you check which git commit you are using?  Anything in the last 2 months should work. @JacobPease is the SD card expert he would have the best insight.

**Comment by PietDelaney** [comment]
I update the wally/cvw repo and tried to update the sub-modules. 
What script are you using to update you wally/cvw repo and the sub-modules?

My last try resulted in the console no longer mentioning Wally. Looks like a LED demo:

 **_********************************************************
********************************************************
**        Avnet/Digilent Arty Evaluation Board        **
**        LEDs and switches GPIO Demonstration        **
********************************************************
********************************************************
**
Choose Task:
BTN0: Print PWM value.
BTN1: 'Cylon' LED display.
BTN2: Scrolling LED display.
BTN3: Return to this menu._**

 
I noticed that the root filesystem on the sdcard has X86 files, not riscV. I'll retry again.
Getting an x86 file-system on the build-root flash is totally bonkers.


I'd like to know how you guys keep you cvw and sub-modules up to date and in sync.
Just want to do a git reset --hard, and the a git pull of everything.

-piet

**Comment by jordancarlin** [comment]
Hi @PietDelaney,

To update the local clone of the repo we usually do a `git pull upstream main` followed by a `git submodule update` (if necessary). The pull from upstream will only work if you set up the upstream remote per the new user setup in the README. 

This won't override any local changes though, so if that is your intention then doing a git reset first might be necessary.

**Comment by PietDelaney** [comment]
I've tried a few re-builds of everything and continue to get the same result:
**_[0.05343] Initializing SPI Controller.                                          
[0.05703] Initializing SD Card in SPI mode.                                     
[0.06125] CMD0: Success, r = 0x01                                               
[0.06435] CMD8: Failed, 0x00F00001AA                                            
[0.06788] Failed at: 0x00000000000017B2_**    

I discussed  fpga/zsbl directory with chat-gpt and tired a few of it's suggestions. 
Even with an virgin SD card the error occurs.  WIthout a SD card it fails at CMD0..

Maybe the author, Jaocb Pease jacob.pease@okstate.edu, could help.

Here is the SDHC cards Im using, 32GByte:

**_https://www.amazon.com/dp/B0D5HD9R18?ref=ppx_yo2ov_dt_b_fed_asin_title&th=1_**

I ordered a few more sdcard adapter to try that.

The Elsevier_3June2024 book instructions for the Wally Makefile in :riscv/cvw/fpga/generator wasnt clear on
editing the Makefile and it looked fine as it was. Was that a mistake.  Im building everything with the stock files in the Wally/cvw git repo.

Do you guys have scripts that you run to build the complete fpga?

This is taking a long time to resolve.

Sigh.

-piet

**Comment by PietDelaney** [comment]
I repeated the build process, did the:

     **git pull upstream main followed by a git submodule update**

then followed the step by step instructions in **_:~/riscv/cvw/fpga/README.md_**

Results are yet again the same:

Welcome to minicom 2.8

OPTIONS: I18n                                                                
Port /dev/ttyUSB1, 18:44:06                                                  
                                                                             
Press CTRL-A Z for help on special keys                                      
                                                                             
 █▀█        █▀█        █▀█        █▀▀                 █ █                    
 █          █ █        █▄▀        █▄▄       ▄▄▄       █ █                    
 █▄█        █▄█        █ █        █▄▄                 ▀▄▀                    
 ____          ____  ____      ___      ___   ____    ___                    
 \   \        /   / /    \    |   |    |   |  \   \  /  /                    
  \   \  __  /   / /      \   |   |    |   |   \   \/  /                     
   \   \/  \/   / /   /\   \  |   |    |   |    \     /                      
    \          / /   ____   \ |   |___ |   |___  |   |                       
     \___/\___/ /___/    \___\|_______||_______| |___|                       
                                                                             
[0.05342] Initializing SPI Controller.                                       
[0.05703] Initializing SD Card in SPI mode.                                  
[0.06125] CMD0: Success, r = 0x01                                            
**[0.06435] CMD8: Failed, 0x00F00001AA
[0.06788] Failed at: 0x00000000000017B2**

This is disappointing. Sigh

**Comment by jordancarlin** [comment]
> The Elsevier_3June2024 book instructions for the Wally Makefile in :riscv/cvw/fpga/generator wasnt clear on
editing the Makefile and it looked fine as it was. Was that a mistake. Im building everything with the stock files in the Wally/cvw git repo.

You are correct. That process (and accompanying section of the textbook) has been updated since that draft. To make the FPGA bitstream for an Arty A7 it should just be a matter of running `make` in the `cvw/fpga/generator` directory. The README in that directory should be up to date.

In terms of debugging the CMD8 failure, could you post the exact commands you are running? I know you said you are following the README, but maybe there is an issue with one of the arguments you are using for one of the scripts.

**Comment by rosethompson** [comment]
Hi Piet,

Just to double check which Pmod port do you have the flash card plugged into?  It should be the right most one labeled JD.

Rose

**Reply by PietDelaney** [reply]
Yea, Im using JD Pmod. I can see it's fine, the 1st test CMD0, only succeeds of a flash card is installed.
The CMD8 code has nothing about the meaning of the errors. No mention of the spec sheet for the registers of the sd.

**Comment by rosethompson** [comment]
I just tested a build from the latest head commit on the main branch and it's working for me.  The commit hash is 528ecfb6. These are the steps I followed.

On my github fork I synced.
Then on my local commit I did...
git checkout main
git pull
cd fpga/generator
make cleanAll
make artya7

After the build I opened the project and programed the device.  I used screen rather than minicom to connect to the UART interface but minicom should work.

screen /dev/ttyUSB2 115200

Edit: I tried with minicom and it works. Added command so I don't forget the syntax.
minicom -D /dev/ttyUSB2 -b 115200

**Comment by rosethompson** [comment]
This is the flash card I'm using, I've also used others so I don't think it will make a difference. 
[Flash card](
https://www.amazon.com/SanDisk-Extreme-UHS-I-Memory-SDSDXXU-064G-GN4IN/dp/B09X7C7NCZ?crid=AYJ5X7AGCD6K&dib=eyJ2IjoiMSJ9.mfDrvgMvVoGH_lGB4ZGqNY99aHQEnG8rd1q7odBesbdv0KRDBXNU9GziUJleCa_4_y5K1MkA7KcU3EuhypJBbqtFfhtkzekTViqcbMBCKSY0V0x7oBOQqbgN-cfiYDizlBHhSu5inieT5L8lf8DvH1vaASAd7pXifmRmevb6Y8lzJIK_8Pu38QaovY7DHqmdf9LdVWkBG_BRbHDENlB_j9qQqgpec-lji9n_-lJ5KaE.HOzcS0mlIVH5gxJ3-VPIYrDI-voow4lY5ucDSicEep4&dib_tag=se&keywords=sandisk+extreme+pro+64gb&qid=1736788414&sprefix=sandisk+extreme+pro+64%2Caps%2C161&sr=8-1)

**Reply by PietDelaney** [reply]
I'vrebeen using 32Gb flash cards. I was worried it might be too large. Yours is 64GB, I ordered one and will try it tomorrow.

**Reply by PietDelaney** [reply, marked as answer]
I tried the 64BG card that your using. It's working much better. I'll try flashing the card next.

********************************************************
********************************************************
**        Avnet/Digilent Arty Evaluation Board        **
**        LEDs and switches GPIO Demonstration        **
********************************************************
********************************************************
**
Choose Task:
BTN0: Print PWM value.
BTN1: 'Cylon' LED display.
BTN2: Scrolling LED display.
BTN3: Return to this menu.

 █▀█        █▀█        █▀█        █▀▀                 █ █
 █          █ █        █▄▀        █▄▄       ▄▄▄       █ █
 █▄█        █▄█        █ █        █▄▄                 ▀▄▀
 ____          ____  ____      ___      ___   ____    ___
 \   \        /   / /    \    |   |    |   |  \   \  /  /
  \   \  __  /   / /      \   |   |    |   |   \   \/  /
   \   \/  \/   / /   /\   \  |   |    |   |    \     /
    \          / /   ____   \ |   |___ |   |___  |   |
     \___/\___/ /___/    \___\|_______||_______| |___|
 
[0.05343] Initializing SPI Controller. 
[0.05703] Initializing SD Card in SPI mode. 
[0.06125] CMD0: Success, r = 0x01 
**[0.06435] CMD8: Success, 0x01000001AA** 
[0.06797] ACMD41: Success, r = 0x00 
[0.29713] New clock frequency: 5000000 
[0.30091] SD card is initialized. 
[0.30408] Getting GPT information. 
          Blocks loaded: 1/1 
[0.31677] Getting partition entries. 
          Blocks loaded: 1/1 
[0.32964] Loading device tree at: 0x8F000000 
          Blocks loaded: 1/1 
[0.34320] Loading OpenSBI at: 0x80000000 
          Blocks loaded: 1/1 
[0.35643] Loading Linux Kernel at: 0x80200000 
          Blocks loaded: 1/1 
[0.37008] Done! Flashing LEDs and jumping to OpenSBI...
